h2o-algos/src/main/java/hex/gam/GamSplines/NBSplinesTypeI.java

```diff
@@ -0,0 +1,110 @@
+package hex.gam.GamSplines;
+
+import java.util.List;
+
+import static hex.gam.GamSplines.NBSplinesUtils.*;
+import static hex.genmodel.algos.gam.GamUtilsISplines.extractKnots;
+
+/**
+ * I implemented the spline described in Section III of doc in JIRA: https://h2oai.atlassian.net/browse/PUBDEV-8398.
+ * Any reference to doc in the code refer to the one here with the http link.
+ * 
+ * The recursive formula in equation 5 is used.  It is implemented as a binary tree with current node with order m
+ * and two child nodes with order m-1.
+ */
+public class NBSplinesTypeI {
+    private static final double EPS = 1e-12;
+    public final int _order;
+    public double[][] _nodeCoeffs;  // expanded polynomial coefficients at current node, section VI of doc
+    private double[] _coeffLeft;    // represent (t-ti) of equation 5
+    private double[] _coeffRite;    // represent (ti+k-t) of equation 5
+    private double[] _knots;        // knot sequence with duplication
+    private double _commonConst;    // k/((k-1)*(ti+k-ti) of equation 5
+    private NBSplinesTypeI _left;   // lower order spline Mi,k-1(t) of equation 5
+    private NBSplinesTypeI _rite;   // lower order spline Mi+1,k-1(t) of equation 5
+    
+    /**
+     * Generate SBSpline Type I in Section III of doc 
+     * 
+     * @param knots : knots that span the whole input range of interest with no duplication
+     * @param order : order of spline
+     * @param basisIndex : basis function index
+     * @param offset : offset added to basis function index, recall lower order splines have index i and i+1 equation 5 of doc
+     * @param numKnotInt : integer denoting knot intervals over which polynomial coefficients are defined
+     */
+    public NBSplinesTypeI(double[] knots, int order, int basisIndex, int offset, int numKnotInt) {
+        _order= order;
+        _knots = extractKnots(offset, order, knots);    // extract knots over basis function is non-zero
+        _nodeCoeffs = new double[numKnotInt][];
+        if (_order == 1) {
+            _commonConst = 0;
+            _coeffLeft = Math.abs(_knots[1]-_knots[0])<EPS?new double[]{0}:new double[]{1.0/(_knots[1]-_knots[0])};
```

Comment:
`_commonConst` and `_coeffLeft` computation could be extracted  to a method for a better readability

Reply:
Good idea.  Done.